File: output/cpp/tests/logical_ops.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>


// Test: Logical operators in Frame handlers
// Tests: &&, ||, !

#include <iostream>
#include <string>
#include <cassert>

class LogicalTestFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    LogicalTestFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class LogicalTestFrameContext {
public:
    LogicalTestFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    LogicalTestFrameContext(LogicalTestFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class LogicalTestCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<LogicalTestFrameEvent> forward_event;
    std::unique_ptr<LogicalTestCompartment> parent_compartment;

    explicit LogicalTestCompartment(const std::string& state) : state(state) {}

    std::unique_ptr<LogicalTestCompartment> clone() const {
        auto c = std::make_unique<LogicalTestCompartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class LogicalTest {
private:
    std::vector<std::unique_ptr<LogicalTestCompartment>> _state_stack;
    std::unique_ptr<LogicalTestCompartment> __compartment;
    std::unique_ptr<LogicalTestCompartment> __next_compartment;
    std::vector<LogicalTestFrameContext> _context_stack;

    void __kernel(LogicalTestFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            LogicalTestFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                LogicalTestFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    LogicalTestFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(LogicalTestFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "Ready") {
            _state_Ready(__e);
        }
    }

    void __transition(std::unique_ptr<LogicalTestCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_Ready(LogicalTestFrameEvent& __e) {
        if (__e._message == "set_values") {
            auto x = std::any_cast<bool>(__e._parameters.at("x"));
            auto y = std::any_cast<bool>(__e._parameters.at("y"));
            a = x;
            b = y;
        } else if (__e._message == "test_and") {
            if (a && b) {
                _context_stack.back()._return = std::any(true);
            } else {
                _context_stack.back()._return = std::any(false);
            }
        } else if (__e._message == "test_not") {
            if (!a) {
                _context_stack.back()._return = std::any(true);
            } else {
                _context_stack.back()._return = std::any(false);
            }
        } else if (__e._message == "test_or") {
            if (a || b) {
                _context_stack.back()._return = std::any(true);
            } else {
                _context_stack.back()._return = std::any(false);
            }
        }
    }

public:
    bool a = true;
    bool b = false;

    LogicalTest() {
        __compartment = std::make_unique<LogicalTestCompartment>("Ready");
        LogicalTestFrameEvent __frame_event("$>");
        LogicalTestFrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    bool test_and() {
        LogicalTestFrameEvent __e("test_and");
        LogicalTestFrameContext __ctx(std::move(__e), std::any(bool()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<bool>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    bool test_or() {
        LogicalTestFrameEvent __e("test_or");
        LogicalTestFrameContext __ctx(std::move(__e), std::any(bool()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<bool>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    bool test_not() {
        LogicalTestFrameEvent __e("test_not");
        LogicalTestFrameContext __ctx(std::move(__e), std::any(bool()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<bool>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    void set_values(bool x, bool y) {
        std::unordered_map<std::string, std::any> __params;
        __params["x"] = x;
        __params["y"] = y;
        LogicalTestFrameEvent __e("set_values", std::move(__params));
        LogicalTestFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }
};

int main() {
    printf("TAP version 14\n");
    printf("1..6\n");

    LogicalTest s;

    // a=true, b=false: true && false = false
    if (!s.test_and()) {
        printf("ok 1 - true && false is false\n");
    } else {
        printf("not ok 1 - true && false is false\n");
    }

    // a=true, b=false: true || false = true
    if (s.test_or()) {
        printf("ok 2 - true || false is true\n");
    } else {
        printf("not ok 2 - true || false is true\n");
    }

    // a=true: !true = false
    if (!s.test_not()) {
        printf("ok 3 - !true is false\n");
    } else {
        printf("not ok 3 - !true is false\n");
    }

    // Change values: a=true, b=true
    s.set_values(true, true);

    // true && true = true
    if (s.test_and()) {
        printf("ok 4 - true && true is true\n");
    } else {
        printf("not ok 4 - true && true is true\n");
    }

    // Change values: a=false, b=false
    s.set_values(false, false);

    // false || false = false
    if (!s.test_or()) {
        printf("ok 5 - false || false is false\n");
    } else {
        printf("not ok 5 - false || false is false\n");
    }

    // !false = true
    if (s.test_not()) {
        printf("ok 6 - !false is true\n");
    } else {
        printf("not ok 6 - !false is true\n");
    }

    return 0;
}

File: output/go/tests/logical_ops.go

package main

import "fmt"

// Test: Logical operators in Frame handlers
// Tests: &&, ||, !

type LogicalTestFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type LogicalTestFrameContext struct {
    _event  LogicalTestFrameEvent
    _return any
    _data   map[string]any
}

type LogicalTestCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *LogicalTestFrameEvent
    parentCompartment *LogicalTestCompartment
}

func newLogicalTestCompartment(state string) *LogicalTestCompartment {
    return &LogicalTestCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *LogicalTestCompartment) copy() *LogicalTestCompartment {
    nc := &LogicalTestCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type LogicalTest struct {
    _state_stack []*LogicalTestCompartment
    __compartment *LogicalTestCompartment
    __next_compartment *LogicalTestCompartment
    _context_stack []LogicalTestFrameContext
    a bool
    b bool
}

func NewLogicalTest() *LogicalTest {
    s := &LogicalTest{}
    s._state_stack = make([]*LogicalTestCompartment, 0)
    s._context_stack = make([]LogicalTestFrameContext, 0)
    s.__compartment = newLogicalTestCompartment("Ready")
    s.__next_compartment = nil
    __frame_event := LogicalTestFrameEvent{_message: "$>", _parameters: nil}
    __ctx := LogicalTestFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *LogicalTest) __kernel(__e *LogicalTestFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &LogicalTestFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &LogicalTestFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &LogicalTestFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *LogicalTest) __router(__e *LogicalTestFrameEvent) {
    switch s.__compartment.state {
    case "Ready":
        s._state_Ready(__e)
    }
}

func (s *LogicalTest) __transition(next *LogicalTestCompartment) {
    s.__next_compartment = next
}

func (s *LogicalTest) TestAnd() bool {
    __e := LogicalTestFrameEvent{_message: "TestAnd"}
    __ctx := LogicalTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result bool
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(bool) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *LogicalTest) TestOr() bool {
    __e := LogicalTestFrameEvent{_message: "TestOr"}
    __ctx := LogicalTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result bool
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(bool) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *LogicalTest) TestNot() bool {
    __e := LogicalTestFrameEvent{_message: "TestNot"}
    __ctx := LogicalTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result bool
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(bool) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *LogicalTest) SetValues(x bool, y bool) {
    __params := map[string]any{
        "x": x,
        "y": y,
    }
    __e := LogicalTestFrameEvent{_message: "SetValues", _parameters: __params}
    __ctx := LogicalTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *LogicalTest) _state_Ready(__e *LogicalTestFrameEvent) {
    if __e._message == "$>" {
        s.a = true
        s.b = false
    } else if __e._message == "SetValues" {
        x := __e._parameters["x"].(bool)
        _ = x
        y := __e._parameters["y"].(bool)
        _ = y
        s.a = x
        s.b = y
    } else if __e._message == "TestAnd" {
        if s.a && s.b {
            s._context_stack[len(s._context_stack)-1]._return = true
        } else {
            s._context_stack[len(s._context_stack)-1]._return = false
        }
    } else if __e._message == "TestNot" {
        if !s.a {
            s._context_stack[len(s._context_stack)-1]._return = true
        } else {
            s._context_stack[len(s._context_stack)-1]._return = false
        }
    } else if __e._message == "TestOr" {
        if s.a || s.b {
            s._context_stack[len(s._context_stack)-1]._return = true
        } else {
            s._context_stack[len(s._context_stack)-1]._return = false
        }
    }
}

func main() {
	fmt.Println("TAP version 14")
	fmt.Println("1..6")

	s := NewLogicalTest()

	// a=true, b=false: true && false = false
	if !s.TestAnd() {
		fmt.Println("ok 1 - true && false is false")
	} else {
		fmt.Println("not ok 1 - true && false is false")
	}

	// a=true, b=false: true || false = true
	if s.TestOr() {
		fmt.Println("ok 2 - true || false is true")
	} else {
		fmt.Println("not ok 2 - true || false is true")
	}

	// a=true: !true = false
	if !s.TestNot() {
		fmt.Println("ok 3 - !true is false")
	} else {
		fmt.Println("not ok 3 - !true is false")
	}

	// Change values: a=true, b=true
	s.SetValues(true, true)

	// true && true = true
	if s.TestAnd() {
		fmt.Println("ok 4 - true && true is true")
	} else {
		fmt.Println("not ok 4 - true && true is true")
	}

	// Change values: a=false, b=false
	s.SetValues(false, false)

	// false || false = false
	if !s.TestOr() {
		fmt.Println("ok 5 - false || false is false")
	} else {
		fmt.Println("not ok 5 - false || false is false")
	}

	// !false = true
	if s.TestNot() {
		fmt.Println("ok 6 - !false is true")
	} else {
		fmt.Println("not ok 6 - !false is true")
	}
}
